EuroPi Script Save/Load State

awonak opened the pull request:
See #102 for design details.

This introduces an interface for saving a script's state variables to be loaded the next time the module is powered back on.

There are 3 supported persistence formats, `string`, `JSON`, and `bytes`.  It is up to the script author to handle the encoding/decoding of their script's state into the persistence formats.

Scripts should be mindful of the size of data being persisted and the frequency of calls to save state (maintainers should look out for this during code review).

New docs:

https://awonak.github.io/EuroPi/generated/europi_script.html

https://github.com/awonak/EuroPi/blob/polyrhythm-save-state/software/contrib/menu.md#saveload-script-state

mjaskula commented:
I was thinking that the menu would call `save_state()` when the quit was requested, but now I'm not sure if this would be helpful or not.

mjaskula commented on software/firmware/europi_script.py:
```diff
@@ -52,23 +52,111 @@ def display_name(cls):
+        return f"saved_state_{self.__class__.__qualname__}.txt"
+
+    def save_state(self, state: str):
+        """Take state in persistence format as a string and write to disk."""
```
This method should get the frequency `..  note::` as well, right?

mjaskula commented on software/firmware/europi_script.py:
```diff
@@ -52,23 +52,111 @@ def display_name(cls):
+            # https://docs.python.org/3/library/struct.html#format-characters
+            format_string = "b?"
+            state = pack(format_string, 2, True)
+            super().save_state_json(state)
```
Should be `save_state_bytes()` in this example.

mjaskula commented on software/firmware/europi_script.py:
```diff
@@ -52,23 +52,111 @@ def display_name(cls):
+
+        ::
+
+            # https://docs.python.org/3/library/struct.html#format-characters
```
The struct stuff is nice, but I wonder if it unnecessarily complicates the example.

Perhaps there should just be a minimal example, followed by a mention that a struct could be useful, and the link.

mjaskula commented on software/firmware/europi_script.py:
```diff
@@ -52,23 +52,111 @@ def display_name(cls):
+        ::
+
+            # Check for a previously saved state. If it exists, return state as a
+            # byte string. If no state is found, an empty string will be returned.
```
> Check for a previously saved state. If it exists, return state as a byte string.

This is redundant with the docstring.

>  If no state is found, an empty string will be returned.

This should be part of the docstring, not a comment in the example.

mjaskula commented on software/firmware/europi_script.py:
```diff
@@ -52,23 +52,111 @@ def display_name(cls):
+        ::
+
+            # Check for a previously saved state. If it exists, return state as a
+            # dict. If no state is found, an empty dictionary will be returned.
```
Same as previous.

mjaskula commented on software/contrib/polyrhythmic_sequencer.py:
```diff
@@ -181,6 +211,46 @@ def _trigger_seq(self, step: int):
+    def load_state(self):
+        """Load state from previous run."""
+        state = super().load_state_bytes()
+        if state == "":
```
i think `if not state:` reads a little nicer, or, even better:

```python
if state:
  self.set_state(state)
```

mjaskula commented on software/contrib/polyrhythmic_sequencer.py:
```diff
@@ -181,6 +211,46 @@ def _trigger_seq(self, step: int):
+    def _ready_to_save(self):
+        if self._dirty:
+            return time.ticks_diff(time.ticks_ms(), self._last_saved) > 1000
+        return False
```
This method could be: 
```python
return self._dirty and time.ticks_diff(time.ticks_ms(), self._last_saved) > 1000
```

mjaskula reviewed (approved):
Looks pretty good to me. Just had a few doc changes and some minor code nitpicks if you want them.

gamecat69 commented:
I have read through the pull request and code example. In general I think it's a good idea and using these pre-defined methods and code examples should simplify fellow coders that want their users to be able to revert to a previous state. For Consequencer, this would be useful to recall previously generated CV sequences.

On the flip-side, coders using this method should probably be also encouraged to consider the following when using these methods:
1. Not saving state too often. You mention this in your docs, but then save state every din.handler() event, which may introduce unwanted latency, especially if attempting to sync with other modules.
2. Knob positions being out of sync with loaded variables. Suggest this level of guidance is added to the docs. For example for Consequencer, I would choose to only save and load the generated CV and not save or load variables that are adjusted using the knobs.
3. A simple method would also be needed to clear a loaded state if the user did not want it. In the case of Consequencer, I believe most users would not want any state saved and prefer to start afresh each time. However, for your poly sequencer, I would imagine users would really like the ability to always recall the previous sequence, but of course be able to clear it quickly with a button/knob combination.

Hope this feedback helps.

awonak commented:
@mjaskula I have updated the docs based on your suggestions, I simplified the helper method docstrings in favor of the more fleshed out example in the README. I also added a new method to provide `last_saved()` duration in ticks since last save since I imagine that's going to be a common use case to reduce the frequency of saves in use cases like in PolyrhythmSeq.

@gamecat69 Thanks for the review and feedback! I updated the docs to make points 1 & 2 more explicit. I didn't add 3 since I imagine that is more of a script design decision that something the base class could handle in a generic way.

mjaskula commented on software/contrib/menu.md:
```diff
@@ -46,6 +46,66 @@ one to import your class and one to add the class to the list. Use the existing
+from europi import oled
+from europi_script import EuroPiScript
+
+class HelloWorld(EuroPiScript):
```
Should we add the save state functionality to the real `hello_world.py`?

mjaskula commented on software/firmware/europi_script.py:
```diff
@@ -32,18 +35,17 @@ def main():  # 3
 
-    def __init__(self) -> None:
-        self._state_filename = f"saved_state_{self.__class__.__qualname__}.txt"
+    _last_saved = 0
```
Why is `last_saved` a class variable?

mjaskula replied:
Yes, but as a class variable it will be shared by all instances, and will not be reset when a new instance is created. The nature of this variable means that it will likely be naturally reset a few seconds after a script starts, so I don't think that under current usage patterns this being a class variable will cause a bug. But I do think that it is an incorrect use of a class variable and not something that we want to make an example of. In the future we may want multiple scripts running in parallel, and this would prevent that.

In addition, I do think that scripts calling `super().__init__()` is a good thing, because (a) it is what a well behaved subclass should do, and (b) it will allow us to more easily add functionality that needs initialization to the base class in the future.

awonak replied:
This is excellent feedback. My intention was to reduce the amount of required steps for script authors when creating a EuroPi script, since we want to try and be as beginner friendly as possible. However, beginners can still start with the one-off main.py script and completely forego the `EuroPiScript` base class, which is arguably intended for script authors with more Python experience.

I'll remove the class variable and add it back in as an instance variable and add `super().__init__()` back into the documentation.

mjaskula commented on software/contrib/menu.md:
```diff
@@ -46,6 +46,66 @@ one to import your class and one to add the class to the list. Use the existing
+            self.enabled = not self.enabled
+            self.save_state()  # 3
+
+    def save_state(self):  # 4
```
Is the intention here to override `EuroPiScript.save_state()`? I don't think that we should recommend this naming, it's going to lead to confusion.

mjaskula commented on software/contrib/polyrhythmic_sequencer.py:
```diff
@@ -181,8 +212,42 @@ def _trigger_seq(self, step: int):
+        if state:
+            self.set_state(state)
+
+    def save_state(self):
```
See note on `save_state()` name in menu.md.

mjaskula commented on software/firmware/europi_script.py:
```diff
@@ -52,23 +54,92 @@ def display_name(cls):
+    def _state_filename(self):
+        return f"saved_state_{self.__class__.__qualname__}.txt"
+
+    def save_state(self, state: str):
```
I'm wondering if this should be `save_state_str()` to avoid the override mentioned in my comment in `menu.md`

mjaskula reviewed (changes requested):
I've added a few more minor comments.

A larger issue I noticed is the override of the `save_state()` and `load_state()` EuroPiScript methods. I don't think that we should recommend this. It works find in the scripts that are using bytes or json format, but won't for str. We should be consistent.

mjaskula commented:
> When dual long pressing the buttons, you return to the menu selection and saved state of the previous script is deleted

This doesn't sound right. Are you referring to this:

```python
def exit_to_menu(self):
    self.remove_state()
    machine.reset() 
```

This code is the deletion of the *bootloader's* state, so that if the module was restarted it would return to the menu rather than the previous running program.

mjaskula commented:
> 3\. A simple method would also be needed to clear a loaded state if the user did not want it. In the case of Consequencer, I believe most users would not want any state saved and prefer to start afresh each time. However, for your poly sequencer, I would imagine users would really like the ability to always recall the previous sequence, but of course be able to clear it quickly with a button/knob combination.

I think that scripts should be smart about the state that they save and it should only be things that the user would want or expect to be saved. I'm curious as to what the situation would be that the user would want to clear a running program's state rather than change it by using the program's UI.

awonak reviewed (commented):
Applied requested fixes and added a few comments.

awonak commented on .vscode/settings.json:
```diff
@@ -0,0 +1,7 @@
+{
```
Oops! Removed.

awonak commented:
This should be ready for final review. Let me know if there are any more areas that can be cleaned up.

Documentation updated: https://awonak.github.io/EuroPi/generated/europi_script.html

mjaskula reviewed (approved):
This looks good to me now. Great work.

roryjamesallen commented on software/contrib/menu.md:
```diff
@@ -46,6 +46,69 @@ one to import your class and one to add the class to the list. Use the existing
+
+When adding save state functionality to your script, there are a few important considerations to keep in mind:
+
+1. Frequency of saves - scripts should only save state to disk when state changes, and should not save too frequently because os write operations are expensive in terms of time. Saving too frequently will affect the performance of a script.
```
I'd capitalise OS as it's an acronym

roryjamesallen commented on software/contrib/menu.md:
```diff
@@ -46,6 +46,69 @@ one to import your class and one to add the class to the list. Use the existing
+When adding save state functionality to your script, there are a few important considerations to keep in mind:
+
+1. Frequency of saves - scripts should only save state to disk when state changes, and should not save too frequently because os write operations are expensive in terms of time. Saving too frequently will affect the performance of a script.
+1. Save state file size - The pico only has about 1MB of free space available so save state storage format is important to keep as minimal as possible.
```
Pico is a brand name so I'd capitalise that too

roryjamesallen commented on software/contrib/menu.md:
````diff
@@ -46,6 +46,69 @@ one to import your class and one to add the class to the list. Use the existing
+```
+
+1. **Initialize base classes** When implementing the `EuroPiScript` base class, its initialization method must be called to initialize its intance variables.
+1. **Call the inherited `EuroPiScript` method `load_state_json()`.** The `EuroPiScript` base class has the method `load_state_json()` to check for a previously saved state. When initializing your script, call `load_X_state()` where `X` is the persistance format of choice. If no state is found, an empty value will be returned.
````
Spelling 'persistance' to persistence

roryjamesallen commented on software/contrib/menu.md:
```diff
@@ -46,6 +46,69 @@ one to import your class and one to add the class to the list. Use the existing
+
+1. **Save state upon state change.** When a state variable changes, call the save state function.
+
+1. **Script save state method.** Provide a helper method to serialize the state variables into a string, JSON, or bytes an call the appropriate save state method.
```
Spelling 'an' to and

roryjamesallen commented on software/firmware/europi_script.py:
```diff
@@ -52,23 +121,99 @@ def display_name(cls):
             pass
+    
+    def last_saved(self):
+        """Return the ticks since last save."""
```
I might add a note that ticks is measured in milliseconds, as this might not be familiar to some people (although if you're using state save functionality you probably will know what it means...) just a suggestion!

roryjamesallen reviewed (changes requested):
Only spelling, everything else looks great!

awonak reviewed (commented):
Typos and edits fixed. Added hooks to save state upon exit to menu.

awonak commented on software/firmware/europi_script.py:
```diff
@@ -126,6 +127,16 @@ def display_name(cls):
+        appropriate save method, such as `save_state_json(state)`. See the
+        class documentation for a full example.
+        """
+        raise NotImplementedError
```
Good call! Done.

mjaskula commented:
I'm happy with this, let's merge it!
